### ISSUE TITLE
fxFlex & fxLayoutGap rendering inconsistency fix proposal

### ISSUE THREAD

**Issue by Alexi996** [ngbracket/ngx-layout#43]
Hello! I've been using FlexLayout for about 3 years now, and I love it. However, there have been some rendering inconsistencies which I want to propose solutions for:

1) fxLayoutGap:
    fxLayoutGap itself can cause several problems.
    Problem 1: When used combined with fxLayout="row wrap", gap is not rendered between element rows (See the image below. Row-wrap is rendered like described in example 1(red), and it's expected to rendered like described in example 2(green)):
![image](https://github.com/ngbracket/ngx-layout/assets/47609114/478769fe-f0d3-49be-af10-8c8191e45349)

  Problem 2)  When used combined with fxLayout="row wrap", and the last rendered row has less elements than previous rows, the last row renders with an extra margin on the left, which makes the row look odd to the previous ones (in some scenarios, this extra-margin is appearing even when all the rendered rows have the same number of elements):

![image](https://github.com/ngbracket/ngx-layout/assets/47609114/1b2ce231-d048-4026-b94b-df8194801d2a)

**Solution proposal:**
-This problem is occurring because fxLayoutGap is using MARGIN for rendering the gap. Solution is to use css **gap** property instead of margin. Additionally, directives for css properties like **row-gap** and **column-gap** can be added to be used in specific cases. **(i've tested this solutions in many scenarios already)**

See the document: https://developer.mozilla.org/en-US/docs/Web/CSS/gap

2) fxFlex rendering inconsistently when using fxLayoutGap and different number of elements in rows:
(Following example was created from a screenshot of one form from real-life project):

![image](https://github.com/ngbracket/ngx-layout/assets/47609114/fd88aab1-d8eb-460b-92db-0af088eabe43)

In Example 1, we can see that one additional fxLayoutGap in the bottom row is creating a rendering inconsistency. One workaround available within current fxLayout capabilitis is to change the HTML nesting and layout, like it's done in Example 2, but a problem arises again if there is another row which needs to be consistent with , and there are less elements (and less fxLayoutGaps).

**Solution proposal:**
1) 90% of those problems are fixed if css **gap** property is used instead of margins **(i've tested this solutions in many scenarios already)**.
2) **SCSS** calculation:
For the most stubborn inconsistencies, I have created several SCSS mixins **(i've tested this solutions in many scenarios already)**.:

//$width is a width of element, $gap is the gap used between elements, and $numberOfElemets is number of elements within a row or column;

@mixin calcWidth($width, $gap, $numberOfElements) {
  width: calc($width - (($gap * ($numberOfElements - 1)) / $numberOfElements));
}

@mixin calcHeight($height, $gap, $numberOfElements) {
  width: calc($height - (($gap * ($numberOfElements - 1)) / $numberOfElements));
}

//$part is a part of row/column which en element needs to take (e.g. 3 for 1/3 of row), $gap is the gap used between elements, and $numberOfElemets is number of elements within a row or column;

@mixin calcWidthPart($part, $gap, $numberOfElements) {
  width: calc((100% / $part) - (($gap * ($numberOfElements - 1)) / $numberOfElements));
}

@mixin calcHeightPart($part, $gap, $numberOfElements) {
  height: calc((100% / $part) - (($gap * ($numberOfElements - 1)) / $numberOfElements));
}

If you need any questions about my proposals, feel free to ask. 
Greetings!

**Comment by DuncanFaulkner** [comment]
Hi, not sure if this is the same or a similar issue, but a fix was put in for layoutgap check this PR https://github.com/ngbracket/ngx-layout/pull/29, it was released in 16.1.3. 
In the meantime, I will try and take a look at your issue, thanks for raising the issue.

**Comment by Alexi996** [comment]
Hi @DuncanFaulkner , thanks for quick response. I just checked PR #29 issue and code, and it is not related to the issue I've posted. Okay, no rush, take your time. Thanks. :)

**Comment by Alexi996** [comment]
Hello, @DuncanFaulkner ! I will need some time study this project before I can start actively participating. But yes, I am willing help :)

**Comment by Alexi996** [comment]
@DuncanFaulkner Can you contact me via email?

**Comment by DuncanFaulkner** [comment]
Yes sure will be next week as I'm tied up with a few things at the moment

**Comment by DuncanFaulkner** [comment]
Fix is up in #110 (targeting the `22.x.x` branch).

**Approach** — implemented @Alexi996's proposal: `fxLayoutGap` now applies the native CSS `gap` property to the flex container instead of adding `margin` to each child. Because `gap` spaces items within a row/column **and between wrapped rows**, this fixes both reported problems:

- **Problem 1** — wrapped rows now get spacing between them.
- **Problem 2** — no more extra leading margin on the last/short row.

It's also direction-independent, so RTL and reversed flows need no special handling. This let a lot of machinery go (MutationObserver, child sorting, layout tracking, the grid negative-margin hack) — net ~−820 lines.

**Verified in the running demo** — rebuilt the demo and drove a browser to the Layout Gap page: a 4-per-row wrapping grid with real gaps between items and between the two wrapped rows, last row flush-left. Confirmed from the live DOM that the container emits `gap: 10px` with zero child margins:

![Layout Gap page](https://raw.githubusercontent.com/ngbracket/ngx-layout/pr-110-verification/layout-gap-verification.png)

**Breaking change to note:** the element must now be a flex container (apply `fxLayout`) for the gap to take effect, and when sizing wrapped children with `fxFlex` you should account for the gap (e.g. `fxFlex="calc(25% - 7.5px)"`). The legacy `" grid"` suffix is still accepted but is now equivalent to a plain gap.

Thanks again @Alexi996 for the detailed report and proposal.
